app/scripts/lib/identity/authentication/index.test.ts

```diff
@@ -0,0 +1,113 @@
+import type MetamaskController from '../../../metamask-controller';
+import { performAutoSignIn } from '.';
+
+type StateOverrides = {
+  participateInMetaMetrics?: boolean;
+  isSignedIn?: boolean;
+  isProfileSyncingEnabled?: boolean;
+  useExternalServices?: boolean;
+  isUnlocked?: boolean;
+};
+
+const arrangeMocks = (stateOverrides?: StateOverrides) => {
+  const performSignInMock = jest.fn();
+  const metamaskControllerInstance = {
+    metaMetricsController: {
+      state: {
+        participateInMetaMetrics: true,
+        ...stateOverrides,
+      },
+    },
+    authenticationController: {
+      state: {
+        isSignedIn: false,
+        ...stateOverrides,
+      },
+      performSignIn: performSignInMock,
+    },
+    userStorageController: {
+      state: {
+        isProfileSyncingEnabled: true,
+        ...stateOverrides,
+      },
+    },
+    preferencesController: {
+      state: {
+        useExternalServices: true,
+        ...stateOverrides,
+      },
+    },
+    keyringController: {
+      state: {
+        isUnlocked: true,
+        ...stateOverrides,
+      },
+    },
+  } as unknown as MetamaskController;
+
+  return { performSignInMock, metamaskControllerInstance };
+};
+
+describe('performAutoSignIn', () => {
+  // @ts-expect-error This is missing from the Mocha type definitions
+  it.each`
+    participateInMetaMetrics | isSignedIn | isProfileSyncingEnabled | useExternalServices | isUnlocked
+    ${false}                 | ${false}   | ${true}                 | ${true}             | ${true}
+    ${true}                  | ${false}   | ${false}                | ${true}             | ${true}
+    ${true}                  | ${false}   | ${true}                 | ${true}             | ${true}
+  `(
+    'should sign the user in when participateInMetaMetrics=$participateInMetaMetrics, isSignedIn=$isSignedIn, isProfileSyncingEnabled=$isProfileSyncingEnabled, useExternalServices=$useExternalServices, isUnlocked=$isUnlocked',
+    ({
+      participateInMetaMetrics,
+      isSignedIn,
+      isProfileSyncingEnabled,
+      useExternalServices,
+      isUnlocked,
+    }: StateOverrides) => {
+      const { performSignInMock, metamaskControllerInstance } = arrangeMocks({
+        participateInMetaMetrics,
+        isSignedIn,
+        isProfileSyncingEnabled,
+        useExternalServices,
+        isUnlocked,
+      });
+
+      performAutoSignIn(metamaskControllerInstance);
+
+      expect(performSignInMock).toHaveBeenCalled();
+    },
+  );
+
+  // @ts-expect-error This is missing from the Mocha type definitions
+  it.each`
```

Comment:
Wicked tests, a bit tough to read in GH side-by-side view, but still pretty cool!

You can convert this into an a 2d array, or an array of objects, but I get the intention and readability behind this.